### ISSUE TITLE
Resources changing and failling

### ISSUE THREAD

**Issue by manzarek123** [OndrejHome/ansible.pcs-modules-2#10]
Sometimes Ansible is trying to make changes to resources when no changes were made to the ansible role/playbook or manually on the cluster.

The first resource will successfully be changed (even if tho there are no changes) but the next ones will fail.
So if I have 10 resources configured I'll have to run 10X Ansible:
On first run it will change the 1st and fail on the 2nd
On second run ti will change the 2nd and fail on the 3rd
....

fatal: [hostname]: FAILED! => {"changed": false, "error": "Error: unable to push cib\nCall cib_replace failed (-205): Update was older than existing configuration\n\n", "failed": true, "msg": "Failed to push updated configuration to cluster using command 'pcs cluster cib-push /tmp/tmpaiRIWh'", "output": ""}

Using latest version of your code,  Ansible 2.4, RHEL7, pacemaker 1.1.15 and corosync 2.4.0

Thanks

**Comment by OndrejHome** [comment]
Hi @manzarek123, Thank you for reporting issue. Are you running the playbook with this behaviour on one node or on multiple nodes of same cluster? Would it be possible to share the playbook and inventory file that you use when this happens?( you can use the Dummy resource if you cannot share original). Can you try running playbook with `--check --diff` and provide the output that will show the changes that module wants to do?

**Comment by OndrejHome** [comment]
To expand on my initial reply here:
I see 2 issues here:
- issue A: changes to resources that seems to be same
- issue B: error when trying to make a change

Having a playbook and inventory file that can reproduce this would be needed. If this is the case where the same change is run from multiple nodes, then `run_once: true` or delegation should be used. 

I can reproduce the issue B when I change the value for existing resource in cluster and then run the `pcs_resource` module on more than one node at a time. This will pop up and is something expected as cluster should be configured only once (one-configuration-per-cluster is enough, no need for one-configuration-per-node).

I will have a look how to document this better for `pcs_resource` or if there is option to serialise it.

**Comment by OndrejHome** [comment]
Hi @manzarek123 , I have recently implemented a change based on Issue #9 that might improve both behaviour of `pcs_resource` module and also the documentation around the usage. Would you be able to let me know if that helps in your case?

Note if there will be no response from you in next 2 weeks I will consider that this is either resolved or no more a problem and will close this issue.

**Comment by manzarek123** [comment]
Hello,
I am using runce_once: true.
I don't really know how to reproduce the issue because it only happens at random time. It never happened when creating the cluster from scratch though. 

I was able to run ansible with --check --diff whith the issue.
Here are the results:
It seems that it is removing some meta_attribute id

Sorry for the late answer :)

TASK [xxx : Create pacemaker FS resource for XXXX] ************************************************************************************************************************
--- before:
+++ after:
@@ -10,5 +10,4 @@
     <op id="xxxxfs-stop-interval-0s" interval="0s" name="stop" timeout="60"/>
     <op id="xxxxfs-monitor-interval-30s" interval="30s" name="monitor" timeout="60s"/>
   </operations>
-  <meta_attributes id="xxxxfs-meta_attributes"/>
 </primitive>

changed: [xxxxxxxxxx.xxx.xxx]

TASK [xxx : Create pacemaker LVM resource for XXXX] ***********************************************************************************************************************
--- before:
+++ after:
@@ -9,5 +9,4 @@
     <op id="xxxxvg-stop-interval-0s" interval="0s" name="stop" timeout="30"/>
     <op id="xxxxvg-monitor-interval-30s" interval="30s" name="monitor" timeout="60s"/>
   </operations>
-  <meta_attributes id="xxxxvg-meta_attributes"/>
 </primitive>

changed: [xxxxxxxxxx.xxx.xxx]

TASK [xxx : Create pacemaker FS resource for XXXX] ************************************************************************************************************************
--- before:
+++ after:
@@ -10,5 +10,4 @@
     <op id="xxxxfs-stop-interval-0s" interval="0s" name="stop" timeout="60"/>
     <op id="xxxxfs-monitor-interval-30s" interval="30s" name="monitor" timeout="60s"/>
   </operations>
-  <meta_attributes id="xxxxfs-meta_attributes"/>
 </primitive>

changed: [xxxxxxxxxx.xxx.xxx]

TASK [xxx : Create Pacemaker MGT VIP for XXXX] ****************************************************************************************************************************
--- before:
+++ after:
@@ -10,5 +10,4 @@
     <op id="xxxx-vip-mgt-stop-interval-0s" interval="0s" name="stop" timeout="20s"/>
     <op id="xxxx-vip-mgt-monitor-interval-30s" interval="30s" name="monitor" timeout="60s"/>
   </operations>
-  <meta_attributes id="xxxx-vip-mgt-meta_attributes"/>
 </primitive>

changed: [xxxxxxxxxx.xxx.xxx]

TASK [xxx : Create Pacemaker XXX VIP for XXXX] ****************************************************************************************************************************
--- before:
+++ after:
@@ -10,5 +10,4 @@
     <op id="xxxx-vip-stop-interval-0s" interval="0s" name="stop" timeout="20s"/>
     <op id="xxxx-vip-monitor-interval-30s" interval="30s" name="monitor" timeout="60s"/>
   </operations>
-  <meta_attributes id="xxxx-vip-meta_attributes"/>
 </primitive>

changed: [xxxxxxxxxx.xxx.xxx]

TASK [xxx : Create Pacemaker MGT VIP for XXXX] ****************************************************************************************************************************
--- before:
+++ after:
@@ -10,5 +10,4 @@
     <op id="xxxx-vip-mgt-stop-interval-0s" interval="0s" name="stop" timeout="20s"/>
     <op id="xxxx-vip-mgt-monitor-interval-30s" interval="30s" name="monitor" timeout="60s"/>
   </operations>
-  <meta_attributes id="xxxx-vip-mgt-meta_attributes"/>
 </primitive>

changed: [xxxxxxxxxx.xxx.xxx]

TASK [xxx : Create Pacemaker XXX VIP for XXXX] ****************************************************************************************************************************
--- before:
+++ after:
@@ -10,5 +10,4 @@
     <op id="xxxx-vip-stop-interval-0s" interval="0s" name="stop" timeout="20s"/>
     <op id="xxxx-vip-monitor-interval-30s" interval="30s" name="monitor" timeout="60s"/>
   </operations>
-  <meta_attributes id="xxxx-vip-meta_attributes"/>
 </primitive>

changed: [xxxxxxxxxx.xxx.xxx]

**Comment by OndrejHome** [comment]
Hi @manzarek123, thank you for more detailed information. I will check this further as the change that is shown there is mostly irrelevant to cluster - there is only indication that empty 'meta' attribute will be removed. I'm making a note on the one entry from pastebin you have provided here just for later reference. I will try this out on the configuration that you have provided. 
One small question here is which version of `pcs` are you using? (`rpm -qa|grep pcs`) 

~~~
TASK [xxx : Create pacemaker LVM resource for XXXX] ***********************************************************************************************************************
--- before:
+++ after:
@@ -9,5 +9,4 @@
     <op id="xxxxvg-stop-interval-0s" interval="0s" name="stop" timeout="30"/>
     <op id="xxxxvg-monitor-interval-30s" interval="30s" name="monitor" timeout="60s"/>
   </operations>
-  <meta_attributes id="xxxxvg-meta_attributes"/>
 </primitive>
 
changed: [xxxxxxxxxx.xxx.xxx]
~~~

**Comment by OndrejHome** [comment]
It looks that I'm able to recreate the "empty" meta attribute by adding and removing 'meta' attributes from the resource with either `pcs resource update xxx meta something=` or `pcs resource meta xxx something=`. In both cases if there was meta attribute `something` with some value it is removed, but it leaves the non-paired XML tag for meta attributes there. Interestingly when I use `pcs resource enable/disable` which essentially creates/deletes meta attribute `target-role` the empty meta tag is removed.

Additional questions for @manzarek123 : 
- 1. Do you use any resource defaults? (`pcs resource defaults`)
- 2. Were the resource that exhibit this problem manipulated in any way that creates and removes meta attributes? (enabled/disable via pcs/webgui, added/remove some meta attributes via `pcs resource meta` or pcs resource update xxx meta yyy=`)

To me this looks more like issue in `pcs` itself at the moment (assuming that `pcs` generates the final XML for CIB). There is no Issue related to this in [ClusterLabs/pcs](https://github.com/ClusterLabs/pcs/issues) which seems as most logical place for that. I will try to have a look at code there to see if it is something easily fixable, if not then lets consider if I can filter out the empty meta tags from comparison in the `pcs_resource` module here.

**Comment by manzarek123** [comment]
I have a repro:
1. Create 2 resources with Ansible, I used LVM and Filesystem.
2. Disable both resources: (this step will generate the meta_attributes)
pcs resource disable myfs
pcs resource disable myvg
3. Enable both resources
pcs resource enable myfs
pcs resource enable myvg
4. Run ansible again, it will remove the meta attribute on the first resource and fail on the second resource.

pcs version:
pcs-0.9.152-10.el7_3.3.x86_64

i.)
resource-stickiness: 100
migration-threshold: INFINITY
ii.) I never used pcs resource meta or webgui.

**Comment by OndrejHome** [comment]
Hi @manzarek123 , I have commit into develop change that should address the empty meta attribute - if code detects that there is empty meta attribute coming from live cluster, it will remove it to have clean comparison. For the error that you mention I'm unable to reproduce that even with your procedure. However you may try to set `force_resource_update=true` for `pcs_resource` module in case you encounter it. That attribute skips the CIB version checking and changes only the resources part of cluster configuration. So unless there is something that changes resources in cluster in paralel it should be OK to use. (resource status changes are in this context not considered to be changes to resources).
Please let me know if these changes helped and if it works now properly for you.

**Comment by OndrejHome** [comment]
Changes are now merged in `master` and available with tag `v13` including on Ansible galaxy. If the problem still persist please let me know,  if not I will close this issue this week.

**Comment by manzarek123** [comment]
It seems to be working fine now :)
Thanks a lot.

Another cool feature could be an option to NOT restart a resource that has been administratively disabled.
For the moment if the resource is disabled it will remove the target-role meta attribute.
Maybe add some "state" attribute with different  value: started, stopped, present (will not restart a disbaled resource), (absent?)

That is just an idea ;)

**Comment by OndrejHome** [comment]
@manzarek123 I will close this issue for now, but if you think that my previous update doesn't answer your idea, feel free to open new issue and we can have a look at it. 
Thank you for your help on identifying the issue with meta attributes here!
